Error in parsing USFM-Grammar

We were using the usfm-grammer for our current POC and we are using npm package. But we were not able to get the ouptut even from the sample code provided in some versions

1.  Issue faced during installing version 3.00-alpha.9
```
npm warn reify invalid or damaged lockfile detected
npm warn reify please re-try this operation once it completes
npm warn reify so that the damage can be corrected, or perform
npm warn reify a fresh install with no lockfile if the problem persists.

``` 

 2.  Got the below error while runs the script.
```
node:internal/modules/cjs/loader:1228
Error: Cannot find module 'tree-sitter-usfm3'

```
3. Fixed by installing
`npm install tree-sitter-usfm3`

4. Next error after running the script
 - script
```
const { USFMParser}  = require('usfm-grammar');
const USFM = '\\id GEN\n\\c 1\n\\p\n\\v 1 In the begining..\\v 2 some more text'
const usfmParser = new USFMParser(USFM);
const USJ = usfmParser.toUSJ()
console.log(USJ);

```

- Error

```
setLanguage.call(this, language);
TypeError: Invalid language object

```
5 . With version  alpha.8, it could only parse small usfm string, and for a large usfm file content, it threw the below error.

![image](https://github.com/user-attachments/assets/0e0671ce-115f-4d9b-92c8-cdaa4c57032e)

- Note:
I have used 42-MRK.usfm for file upload.

--- a/node-usfm-parser/package.json
+++ b/node-usfm-parser/package.json
@@ -27,7 +27,7 @@
   ],
   "dependencies": {
     "tree-sitter": "0.21.1",
-    "tree-sitter-usfm3": "file:../tree-sitter-usfm3",
+    "tree-sitter-usfm3": "3.0.0-beta.9",
     "xmldom": "^0.6.0",
     "xpath": "^0.0.34"
   },

--- a/node-usfm-parser/src/usjGenerator.js
+++ b/node-usfm-parser/src/usjGenerator.js
@@ -493,7 +493,6 @@ class USJGenerator {
       case "text":
         let textVal = this.usfm
           .substring(node.startIndex, node.endIndex);
-          // .trimStart();
         textVal = textVal.replace("~", " ")
         if (textVal !== "") {
           parentJsonObj.content.push(textVal);

--- a/node-usfm-parser/src/usjGenerator.js
+++ b/node-usfm-parser/src/usjGenerator.js
@@ -492,8 +492,8 @@ class USJGenerator {
         break;
       case "text":
         let textVal = this.usfm
-          .substring(node.startIndex, node.endIndex)
-          .trim();
+          .substring(node.startIndex, node.endIndex);
+          // .trimStart();
         textVal = textVal.replace("~", " ")
         if (textVal !== "") {
           parentJsonObj.content.push(textVal);

--- a/node-usfm-parser/src/usxGenerator.js
+++ b/node-usfm-parser/src/usxGenerator.js
@@ -541,7 +541,7 @@ class USXGenerator {
         } else if (node.type.endsWith("Attribute")) {
             this.node2UsxAttrib(node, parentXmlNode);
         } else if (node.type === "text") {
-            let textVal = this.usfm.slice(node.startIndex, node.endIndex).trim();
+            let textVal = this.usfm.slice(node.startIndex, node.endIndex);
             textVal = textVal.replace("~", " ")
             const textNode = parentXmlNode.ownerDocument.createTextNode(textVal);
             parentXmlNode.appendChild(textNode);

--- a/node-usfm-parser/src/usxGenerator.js
+++ b/node-usfm-parser/src/usxGenerator.js
@@ -543,8 +543,10 @@ class USXGenerator {
         } else if (node.type === "text") {
             let textVal = this.usfm.slice(node.startIndex, node.endIndex);
             textVal = textVal.replace("~", " ")
-            const textNode = parentXmlNode.ownerDocument.createTextNode(textVal);
-            parentXmlNode.appendChild(textNode);
+            if (textVal !== "") {
+                const textNode = parentXmlNode.ownerDocument.createTextNode(textVal);
+                parentXmlNode.appendChild(textNode);
+            }
         } else if (["table", "tr"].concat(TABLE_CELL_MARKERS).includes(node.type)) {
             this.node2UsxTable(node, parentXmlNode);
         } else if (node.type === "milestone" || node.type === "zNameSpace") {

--- a/node-usfm-parser/test/test_usj_conversion.js
+++ b/node-usfm-parser/test/test_usj_conversion.js
@@ -187,6 +187,7 @@ function stripTextValue(usjObj) {
             } else {
                 stripTextValue(item);  // Recursively handle nested objects
             }
+        usjObj['content'] = usjObj['content'].filter(item => item === "")
         });
     }
 }

--- a/py-usfm-parser/src/usfm_grammar/usj_generator.py
+++ b/py-usfm-parser/src/usfm_grammar/usj_generator.py
@@ -319,7 +319,7 @@ class USJGenerator:
             self.node_2_usj_attrib(node, parent_json_obj)
         elif node.type == 'text':
             text_val = self.usfm[node.start_byte:node.end_byte].decode(
-                        'utf-8').lstrip()
+                        'utf-8') #.lstrip()
             if text_val != "":
                 parent_json_obj['content'].append(text_val)
         elif node.type in ["table", "tr"]+ self.TABLE_CELL_MARKERS:

--- a/py-usfm-parser/src/usfm_grammar/usx_generator.py
+++ b/py-usfm-parser/src/usfm_grammar/usx_generator.py
@@ -356,11 +356,12 @@ class USXGenerator:
             self.node_2_usx_attrib(node, parent_xml_node)
         elif node.type == 'text':
             text_val = self.usfm[node.start_byte:node.end_byte].decode('utf-8')
-            siblings = parent_xml_node.findall("./*")
-            if len(siblings) > 0:
-                siblings[-1].tail = text_val
-            else:
-                parent_xml_node.text = text_val
+            if text_val != "":
+                siblings = parent_xml_node.findall("./*")
+                if len(siblings) > 0:
+                    siblings[-1].tail = text_val
+                else:
+                    parent_xml_node.text = text_val
         elif node.type in ["table", "tr"]+ self.TABLE_CELL_MARKERS:
             self.node_2_usx_table(node, parent_xml_node)
         elif  node.type == "milestone":

--- a/py-usfm-parser/src/usfm_grammar/usx_generator.py
+++ b/py-usfm-parser/src/usfm_grammar/usx_generator.py
@@ -355,7 +355,7 @@ class USXGenerator:
         elif node.type.endswith("Attribute"):
             self.node_2_usx_attrib(node, parent_xml_node)
         elif node.type == 'text':
-            text_val = self.usfm[node.start_byte:node.end_byte].decode('utf-8').strip()
+            text_val = self.usfm[node.start_byte:node.end_byte].decode('utf-8')
             siblings = parent_xml_node.findall("./*")
             if len(siblings) > 0:
                 siblings[-1].tail = text_val

--- a/py-usfm-parser/tests/test_json_conversion.py
+++ b/py-usfm-parser/tests/test_json_conversion.py
@@ -155,6 +155,7 @@ def strip_text_value(usj_dict):
                 usj_dict['content'][i] = item.strip()
                 continue
             strip_text_value(item)
+        usj_dict['content'] = list(filter(lambda x: x == '', usj_dict['content']))
 
 
 def strip_default_attrib_value(usj_dict):

--- a/web-usfm-parser/src/usjGenerator.js
+++ b/web-usfm-parser/src/usjGenerator.js
@@ -492,8 +492,8 @@ class USJGenerator {
         break;
       case "text":
         let textVal = this.usfm
-          .substring(node.startIndex, node.endIndex)
-          .trim();
+          .substring(node.startIndex, node.endIndex);
+          // .trimStart();
         textVal = textVal.replace("~", " ")
         if (textVal !== "") {
           parentJsonObj.content.push(textVal);

--- a/web-usfm-parser/src/usxGenerator.js
+++ b/web-usfm-parser/src/usxGenerator.js
@@ -546,8 +546,10 @@ class USXGenerator {
         } else if (node.type === "text") {
             let textVal = this.usfm.slice(node.startIndex, node.endIndex);
             textVal = textVal.replace("~", " ")
-            const textNode = parentXmlNode.ownerDocument.createTextNode(textVal);
-            parentXmlNode.appendChild(textNode);
+            if (textVal !== "") {
+                const textNode = parentXmlNode.ownerDocument.createTextNode(textVal);
+                parentXmlNode.appendChild(textNode);
+            }
         } else if (["table", "tr"].concat(TABLE_CELL_MARKERS).includes(node.type)) {
             this.node2UsxTable(node, parentXmlNode);
         } else if (node.type === "milestone" || node.type === "zNameSpace") {

--- a/web-usfm-parser/src/usxGenerator.js
+++ b/web-usfm-parser/src/usxGenerator.js
@@ -544,7 +544,7 @@ class USXGenerator {
         } else if (node.type.endsWith("Attribute")) {
             this.node2UsxAttrib(node, parentXmlNode);
         } else if (node.type === "text") {
-            let textVal = this.usfm.slice(node.startIndex, node.endIndex).trim();
+            let textVal = this.usfm.slice(node.startIndex, node.endIndex);
             textVal = textVal.replace("~", " ")
             const textNode = parentXmlNode.ownerDocument.createTextNode(textVal);
             parentXmlNode.appendChild(textNode);

--- a/web-usfm-parser/test/test_usj_conversion.js
+++ b/web-usfm-parser/test/test_usj_conversion.js
@@ -205,6 +205,9 @@ function removeNewlinesInText(usjDict) {
                 removeNewlinesInText(item);  // Recursively handle nested dictionaries
             }
         });
+        // there will be difference in number of white space only text snippets
+        usjObj['content'] = usjObj['content'].filter(item => item === "")
+
     }
 }
 

--- a/web-usfm-parser/test/test_usj_conversion.js
+++ b/web-usfm-parser/test/test_usj_conversion.js
@@ -206,7 +206,7 @@ function removeNewlinesInText(usjDict) {
             }
         });
         // there will be difference in number of white space only text snippets
-        usjObj['content'] = usjObj['content'].filter(item => item === "")
+        usjDict['content'] = usjDict['content'].filter(item => item === "")
 
     }
 }

